### PR TITLE
[SPARK-9462][SQL] Initialize nondeterministic expressions in code gen fallback mode.

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

**Comment by rxin** [comment]
@mengxr  I think this fixes the problem you ran into.

cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GeneratePredicate.scala [hunk 1]
```diff
@@ -62,6 +62,11 @@ object GeneratePredicate extends CodeGenerator[Expression, (InternalRow) => Bool
 
     logDebug(s"Generated predicate '$predicate':\n${CodeFormatter.format(code)}")
 
+    ctx.references.foreach {
+      case n: Nondeterministic => n.setInitialValues()
+      case _ =>
+    }
+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GeneratePredicate.scala]: it should be
```scala
ctx.references.foreach(_.foreach {
  case n: Nondeterministic => n.setInitialValues()
  case _ =>
})
```

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GeneratePredicate.scala]: and why not do initiliaztion in `CodegenFallback`?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GeneratePredicate.scala]: ah yes

**Comment by rxin** [comment]
@cloud-fan updated.

**Comment by cloud-fan** [comment]
lgtm

**Pull request closed without merging**